lego/apps/events/models.py

```diff
@@ -930,26 +930,46 @@ def set_presence(self, presence: constants.PRESENCE_CHOICES) -> None:
         """Wrap this method in a transaction"""
         if presence not in constants.PRESENCE_CHOICES:
             raise ValueError("Illegal presence choice")
+
         self.presence = presence
         self.handle_user_penalty(presence)
         self.save()
 
+    def delete_penalties_for_event(self) -> None:
+        for penalty in self.user.penalties.filter(source_event=self.event):
+            penalty.delete()
+
     def handle_user_penalty(self, presence: constants.PRESENCE_CHOICES) -> None:
+        """
+        Previous penalties related to the event are deleted since the
+        newest presence is the only one that matters
+        """
+
```

Comment:
Is the logic here that if you get penalties for not paying on time you will no longer have those penalties if you also didn't show up to the event? I would've thought that they rather stack on top of each other - so that each penalty-worthy action got you a new penalty? Or is that not how we want to do it?

Reply:
> I would like some input as to how this will affect other penalties, e.g. given due to late payments (or other things?), as they will also be removed if the presence is changed to something worthy of a penalty. Should we consider to introduce "penalty types" stored on the record, e.g. "presence", "payment", "other"?

Reply:
Yes, they should definitely stack. Just not sure how this should be implemented.

Reply:
Ah right.. I thought the stacking functionality was already implemented in the redo Arash did last year to make penalties last for x amount of days instead of adding onto each other - or maybe I remember that wrong

Reply:
so do you have some input? 😄 If we add penalty types, we can flag them as being either presence or payment related, making it easy to delete the right penalties

Reply:
![image](https://github.com/webkom/lego/assets/13599770/ce0f6d55-6d4d-47e2-b6c4-22ab19278950)

Don't have too much rn - but I'll look through the penalty implementation some time after I get back

Initial thoughts were that not too much had to be changed, as penalties already stack (or so I thought), so you could just add the penalty without deleting the others, but by your question it appears that is not the case so I'll have to review it

Reply:
haha no worries - have fun!

> Initial thoughts were that not too much had to be changed, as penalties already stack (or so I thought), so you could just add the penalty without deleting the others, but by your question it appears that is not the case so I'll have to review it

I agree that they should stack, most of the time. If both penalties are e.g. related to the event presence, I don't believe they should stack. This logic also works as a safety net by allowing missclicks

Reply:
Ah yeah, I very much agree with that! In that way it sounds like a good idea to add types like you suggested, so you could clear the penalties for attendance status for the event when it's changed and keep the penalties for late payment, etc.

Reply:
Just making sure that my input is clearly stated now that I'm back on my computer;

I've been thinking about it and I agree that your proposal is probably the best idea (adding penalty types). So then users would get penalties for late payments and such, and when their presence is changed, only the presence-penalties are modified. This would also be a nice way to remove late penalties from users if they get their status changed to present at a later point due to some mistake (or misclicks as you said)

Realised that part of my confusion came from thinking that this pr had been merged ages ago - which it obviously hasn't https://github.com/webkom/lego/pull/3349